Pull in LessImproved

braver opened the pull request:
Sublime Text 3 now supports a new syntax format that's more powerful and easier to maintain. Versions of ST that don't support the new format will still use the old files. I would propose keeping them around as is and freezing them, with improvements and fixes only in the new format. 

In my fork I simplified a couple of things to keep maintenance cost as low as possible. A list of what changed and why:
- Semi-colon auto-insert removed. And option that defaults to off isn't very discoverable. Also, I think this package should focus on syntax and autocompletion, all other tricks are best kept in user space or in a separate package.
- Language name changed from LESS to Less because that's how they write it.
- LESS.sublime-syntax file converted to the new format and fixing a bunch of things. Most notably it now incorporates keyword lists I once collected for the [Less package for Atom](https://github.com/atom/language-less/pull/42), which fixes #77 and similar.
- I removed the JSON and the make setup because it's much easier to edit the new file directly and it makes it easier to contribute.
- less_completions.py has been updated with new completions from the core CSS package.
- Removed some whitespace from the snippets.
- Removed the main menu entry as it just adds clutter.
- Cleaned up all the outdated information from the readme.
- Removed the example less file because it's not so useful and needs maintenance.

I think this sets up the package to be more reliable and easily maintained going forward. It's always tricky updating a package that's so popular, people don't like change. So perhaps there should be a plan to migrate people who rely on semi-colon auto-insertion. 

@danro I'd like to hear if you would consider pulling this in. If you don't that's totally fine too.

Fixes: #73 #69 #77 #72 #67 #76. Probably also fixes #75 #78 (whatever the reason they've got issues anyway).

danro commented:
Thanks for the work here @braver. Initial thoughts:

- Semicolon insertion seems no different than autocomplete or bracket matching to me. Some users depend on it, so I think it should stay and remain optional. Why remove it and force users to migrate? Was it causing you trouble?

- I don't mind maintaining `tests.less` as it serves as the only way currently to check for regressions. It's a manual, visual test.. but it works. I noticed Sublime is now doing this in their default packages (unsure if theirs is automated).

Lastly, I just noticed in dev build `3107` that Sublime now has a mysterious `Less` syntax of their own. Are you seeing this as well?

braver commented:
> Semicolon insertion seems no different than autocomplete or bracket matching to me. Some users depend on it, so I think it should stay and remain optional. Why remove it and force users to migrate? Was it causing you trouble?

Good points all. I guess it's mostly something I never used and didn't want to support in my own package. But you make a good case for keeping it, and if it works, then why not. I'll put it back in.

> I don't mind maintaining tests.less as it serves as the only way currently to check for regressions. It's a manual, visual test.. but it works. I noticed Sublime is now doing this in their default packages (unsure if theirs is automated).

There is now a way to have an automated test. Looking at their [css test file](https://github.com/sublimehq/Packages/blob/master/CSS/syntax_test_css.css) there is apparently some syntax you can add and Syntax Test build system you can run. I haven't used it yet, but it should certainly be useful. The test file was giving some false negatives, so it has to go. I'll dig up a better test file from somewhere and look into adding some real tests.

> Lastly, I just noticed in dev build 3107 that Sublime now has a mysterious Less syntax of their own. Are you seeing this as well?

I'm not seeing a Less syntax in 3107? There's also nothing in the [packages repo](https://github.com/sublimehq/Packages).

braver commented:
@danro I've restored the semi-colon auto-insertion feature and the proper menu entries. Additionally, I updated the readme to include a message about the focus being on the `.sublime-syntax` file for moving forward and fixing bugs, while keeping the old `.tmLanguage` file around for compatibility. 

About tests: I added a small less file with a single assertion. This will now be tested whenever the build system "Syntax Tests - All Syntaxes" is run. Also, with the file open, you can run the basic "Syntax Test" and it will tell you what has been broken. The test can be filled by hitting CMD-ALT-P and copying the existing scope (assuming it's correct of course). I will be building on this test, and all future pull requests should have a matching entry in the file to assert we don't regress.

Anything else you would want me to look at? If you feel good about where this is going I'll ask @FichteFoll to revert https://github.com/wbond/package_control_channel/pull/5223.

danro commented:
Thanks @braver, still need to test this out. I did figure out the double syntax issue, was being caused by the material theme: https://github.com/equinusocio/material-theme/issues/713

braver commented:
Oh man, are they packaging syntax files with their themes? That'll cause issues... Anyway, let me know what you find and if there's anything you want me to do.

braver commented:
For anyone who's interested in testing these changes, please check out my [LessImproved](https://packagecontrol.io/packages/LessImproved) package. Would love to get anyone's feedback!

braver commented:
@danro any chance of getting some action or feedback on this? I'm still totally motivated to make this merge happen and would love to be able to support this package going forward. If there is anything you need me to verify or test or change or whatever, let me know.

braver commented:
Now that 3.0 is out, the backwards compatibility isn’t really useful anymore. Want me to revise the PR without it?

danro commented:
@braver sure, I'd like to see this package stay current.. thanks for being persistent! 👍

braver commented:
No problem. It’ll be next week before I’m in range of a laptop again, but I’ll push an update and let you know 👍🏻

braver commented:
@danro I merged in all work I've done on my fork and took out the old tmLanguage file. Mostly lots of fine-tuning, more tests, and better adherence to the [ST3.0 scope naming guidelines](https://www.sublimetext.com/docs/3/scope_naming.html). Means it's no longer compatible with ST2, and more up to date with things like custom properties (aka css variables).

braver commented:
@danro to properly merge this, we'd have to move to tag-based releases first so we can do messaging to end-users. If you add me as a collaborator I'll happily take care of the updates in package control, creating releases and managing open (and any new) issues and pull requests and migrating LessImproved users.

braver commented:
@danro if you want to move forward on this please add me as a collaborator and I'll carefully roll this out to users. If not that's ok too, but then let's just close this PR.

danro commented:
Invite sent. Thank you again for all the hard work, I'm sure the LESS users will appreciate it. 👏

braver commented:
Thanks a bunch!

Zzortell commented:
Waouh @braver, your work and persistence are so impressive! Well done!
Signed: an anonymous random end-user of LESS-sublime :)
